Sacha

XachaB opened the pull request:
Import from [old lexibank](https://github.com/glottobank/lexibank-data/tree/master/datasets/tukanoaligned), as discussed in issue lexibank/lexibank#149. 

Also pinging @tresoldi and @thiagochacon in case they want to have a look.

Commit:
Init with cldfbench new

Commit:
Importing from https://github.com/glottobank/lexibank-data/tree/master/datasets/tukanoaligned

Commit:
Init contributor table

Commit:
Updates naming conventions

Commit:
Implements download and makecldf

Commit:
Adds note from original repo

Commit:
Adds downloaded and generated files

Commit:
Corrects a typo in 'Linguistics'

Commit:
Updates and formats orthography profile

Commit:
Corrects title by removing final dot

Commit:
Uses concept list from concepticon, not csv file

Commit:
Computes segments rather than passing them from raw

Commit:
Corrects language IDs to conventional IDs

Commit:
Adds reference for the cognate sets

Commit:
Ignores libreoffice temporary files

Commit:
Updates generated files

Commit:
Solves BIPA and CLTS errors in orthography profile

Commit:
Updates generated files

Commit:
Adds tests: check number of items in each table.

tresoldi commented:
Great work, @XachaB !

LinguList commented on .gitignore:
```diff
@@ -1,3 +1,6 @@
+bank stuff
+cldf/.transcription-report.json
```
we do not ignore the transcription report usually, but upload it, right @xrotwang ?

XachaB replied:
This line was auto-generated, I will change it, but maybe it is possible to change `cldfbench new` so that this is not ignored by default anymore ?

LinguList commented on .zenodo.json:
```diff
@@ -0,0 +1,99 @@
+            "type": "DataCurator"
+        }
+    ],
+    "communities": [
```
Which lexibank version do you use? I know we have this bug, by which the "communities" will be expanded each time, rather than reduced.

LinguList replied:
But I thought we had a solution already.

XachaB replied:
it is pylexibank==2.7.0:
https://github.com/lexibank/chacontukanoan/blob/a6e41d4fc5efde65172b9b4008c794303eee47b2/cldf/requirements.txt#L30

chrzyki replied:
Fix is already in pylexibank but we haven't done a release since then.

LinguList commented on README.md:
```diff
@@ -0,0 +1,50 @@
+
+> T. Chacon. (2014). A revised proposal of Proto-Tukanoan consonants and Tukanoan family classification. Journal of American Linguistics 80.3, pp. 275–322. doi: https://doi.org/10.1086/676393
+
+This dataset is licensed under a GPL-3.0 license
```
We should change the license to CCBY

XachaB replied:
ok, will do.

LinguList commented on etc/orthography.tsv:
```diff
@@ -0,0 +1,89 @@
+ũ	ũ	U+0169	91
++	+	U+002b	11
+-	+		
+_	_	U+005f	4156
```
We now render the `_` as `+` always, as we say that we cannot meaningfully segment words affixes, etc., so we just say: one segmenter of meaning units.

XachaB replied:
ok

LinguList commented on lexibank_chacontukanoan.py:
```diff
@@ -0,0 +1,109 @@
+        brackets={"(": ")"},  # characters that function as brackets
+        separators=";/,",  # characters that split forms e.g. "a, b".
+        missing_data=('?', '-'),  # characters that denote missing data.
+        strip_inside_brackets=True
```
to be on the save side, please add "first_form_only=True"

LinguList replied:
wait: if we add the form, it means, we do not have a formspec, so the form-spec won't do anything, so you can just leave it blank. The whole "form_spec" line.

XachaB replied:
ok

LinguList reviewed (approved):
Good first try. There are two things missing, which we didn't fix last time:

1. alignments should be added (we have them nicely already)
2. tokens are given in tokenized form already, so it may be useful to check if one can add those as well (they should be converted to our target alphabet, but ```add_form_with_segments``` is worth being tested here.

XachaB commented:
Thanks for the review.

Regarding 1., I thought they were automated alignments, and that Lexibank only recorded manual alignments ?
for 2., I have tried using `add_form_with_segments` in a previous version, but if I do so, the orthography profile is not used anymore, which causes issues. Do you know how to solve this ?

Commit:
Changes all separators to +

Commit:
Removes unused form spec

Commit:
Adds transcription report

Commit:
Changes licence to CCBY

Commit:
Adds generated files and formatted profile

LinguList commented:
Alignments are done by Herr Thiago Costa Chacon himself! That's the main idea of the edictor here.

XachaB commented:
Oh, I see ! Great, I'm on it, then :)

LinguList commented:
As to the orthoprofile. I was just thinking about this. We have a couple of similar situations, but no really good solution, so I'd like to use this maybe to think of a principled way (since we also need to adjust the alignments). 

BTW: if you look at the edictor app, you can also inspect how the data look like: https://digling.org/edictor/?file=tukano

LinguList commented:
@tresoldi, may I ask you to think about the following: can we find a way to have a segmented string and then re-segment it with a given orthoprofile? We could artificially add "_" for all spaces, and then run the profile on it, turning the "_" (or the other symbol) to NULL? In general, it would be good if we have some good solution. Then, @XachaB could still add forms with `add_form` but would pass the segmented data, the TOKENS in our edictor application.

XachaB commented:
The alignments too would need to be able to go through the orthography profile, right ? except that any rule which applied to the gap "-" needs to be ignored, otherwise we would completely strip the alignment data.

Could someone point me to a repository where alignments are provided ? I can not find the syntax to add them.

Commit:
Changes one more separator to +

Commit:
Adds generated files and formatted profile

LinguList commented:
you add teh alignment to the cognate set:


```add_cognate(

    Alignment=X

```

tresoldi commented:
@LinguList maybe we could use the [`segments` library](https://github.com/cldf/segments)? The profile object is technically already there, so it should be a matter of wrapping the string/list of segments when passing.

An alternative could be to apply a profile (if available) even when calling `add_form_with_segments()`, i.e., having `cldfbench` take care of it. I suppose this solution is more coherent in the big scheme of things (assuming @xrotwang agrees), but it involves new versions/releases and might even break some stuff we have.

Commit:
Adds expert alignments + regenerates files

XachaB commented:
Maybe this could be an option to `add_form_with_segments` ? something like `add_form_with_segments(..., apply_profile=True)` to recompute alignments, otherwise by default do not (which would ensure backward compatibility) ?

LinguList commented:
Yes, I think using segments is probably the best. We can later think
about bigger solutions. The workaround is as this:

1. use add_form

2. instead `Form=wl[idx, 'ipa']`, write `Form='.'.join(wl[idx,
'tokens'])`, and add `.` to the profile, and characterise it as NULL

I think that should do the trick, right?

LinguList commented:
Maybe worth to add an issue to lexibank/lexibank. It IS useful to modify
the segmented forms, as they are often flawed, and I made a lot of
workarounds so far, which is not the best idea.

XachaB commented:
> Yes, I think using segments is probably the best. We can later think about bigger solutions. The workaround is as this: 1. use add_form 2. instead `Form=wl[idx, 'ipa']`, write `Form='.'.join(wl[idx, 'tokens'])`, and add `.` to the profile, and characterise it as NULL I think that should do the trick, right?

This works for the forms, but the alignments will not benefit from it. Should we also do something to the alignments ?

Commit:
Uses workaround to re-segment from TOKENS

LinguList commented:
For the alignments, one needs to USE the tokenizer directly.

Commit:
Fixes bug in workaround for tokens

Commit:
Re-tokenized alignments, adds more thorough testing of alignments

Commit:
Correcting the alignment for form Barasano-16-buritipalm-1

XachaB commented:
Ok, so I implemented the re-tokenization of alignments. However, I was a bit worried about the consequences that could have (since if we split any aligned segment further, this could break the alignments). To make sure, I implemented more thorough tests for the alignments, checking that:

- All the aligned cognates have the same number of columns
- And the aligned words have the same segments as the segmented form (from TOKENS)

Unfortunately, this second condition is often false. Here are some examples:

The entry for buriti palm in BAS:
![image](https://user-images.githubusercontent.com/5927813/88304401-81d6ec80-cd08-11ea-8833-e9e44348bde8.png)
which I "corrected" to "r̃ e - e", but I am not sure it is the right choice anymore, since both "ẽ" and "ɾ̃" are valid segments.

Several entries for snake: (look at KOR, MAI, SEK, SIO):
![image](https://user-images.githubusercontent.com/5927813/88304059-0bd28580-cd08-11ea-9fa0-f3f91842b6be.png)

But this is also the case for entries for nose, hot, etc.
These can be found by running the current test suite.

I see several possibilities:
- Maybe my assumption was wrong, and it is intended that alignments can have different segments compared to the tokens/segmented form
- Or there are pervasive errors in the alignments. In which case, how can I choose how to resolve it ?

LinguList commented:
If the second condition is false, the alignment deserves preference. Edictor now keeps the forms synced, it did not by then. 

Luckily, we can adjust this:

1. take alignment
2. alignment -> segments/tokens: ```lambda x: [y for y in x if y != '-']```
3. and then use the segments, etc.?

XachaB commented:
Ok, thanks. Is it because the alignments were edited by hand ?

LinguList commented:
yes, and they are from 2015... early unpublished version of edictor, and
the like.

Commit:
When alignments and token conflict, picks the alignment.

Commit:
Orthography profile with two rules for e+tilde and precombined e+tilde

XachaB commented:
I still have an issue, on which I am stuck. Some forms in the data have "ẽ" (precombined) and some have "ẽ" (latin e + combining tilde). Both need to be converted to the normalized version. I wrote two rules:

https://github.com/lexibank/chacontukanoan/blob/5396444bb7b2db9d66c5b92e1f9746f26cce3014/etc/orthography.tsv#L18-L19

However, whatever I do, cldfbench keeps issuing a warning:

~~~
WARNING:segments.profile:line 19:duplicate grapheme in profile: ẽ
~~~

Doing `cldfbench lexibank.format_profile --sort lexibank_chacontukanoan.py`, even without `--trim` always removes the line with the non-precombined item. 

And in any cases, even without formatting the profile first, the non pre-combined version leads to an error in parsing, which is silently swallowed and generates a character `�`. The form `ɾ ẽ e`  becomes `ɾ e � e`. 

I have tried everything I could think of. Obviously, I could just replace this one form (it is the only one with the e+combining tilde), but it seems to me that the orthography profile *should* be able to deal with such a simple replacement. 

What am I missing ?

LinguList commented:
We normalize data before converting, and since this falls in unicode
normalization range, it would be captured. However, if you do not
normalize your input sequence, it will fail. I think, LingPy does not
normalize (or the other way?!?, but we didn't have this issue
otherwise), so you should only need one line. If not, check what happens
if you explicitly normalize:

```
import unicodedata
normalize = lambda x: unicodedata.normalize("NFD", x) # or NFC
```

LinguList commented on lexibank_chacontukanoan.py:
```diff
@@ -0,0 +1,145 @@
+        """
+        data = self.raw_dir.read_csv('tukano.csv', dicts=True)
+        ortho_profile = self.orthography_profile_dict[None]
+        tokenizer = Tokenizer(profile=ortho_profile)
```
Yes, here it is: since you use the tokenizer explicitly, and not as part of the class, as it is loaded anyway, you lose normalization. Our tokenizer is a bit more explicit. @chrzyki or @tresoldi do you remember how I get the "normal" tokenizer from within the args/class of a dataset? If you don't know, never mind, and I check in the pylexibank source code tomorrow, but basically, @XachaB, you don't have to do this explicitly, as the tokenizer is an attribute of the dataset class.

tresoldi replied:
This is actually the solution we came up with, because using directly the class tokenizer requires to pass an empty dictionary to the first argument (the `item`), as it expects the complete CLDF table row. It could be done, of course, but the code looked a bit more cryptic in our view.

The `None` indexing for the `orthography_profile_dict` (necessary since the multiple profile support) might be a bit obscure if one is not used to `pylexibank`, but I suppose it is a better solution and one could always include a short comment explaining the rationale...

XachaB replied:
@LinguList: Ok, this clarifies my problem.

There is a `self.tokenizer` in the instance, which is defined as a lazy property. But contrary to what the name suggests, it is not directly a `Tokenizer` instance, but instead a function wrapping it and expecting an `item` first argument (a dict). As @tresoldi  said, passing it empty sounded like a hack and we thought probably not the intended usage. 

Depending on what you think is more coherent with the rest of the code base, I can either use `self.tokenizer`, or I can simply normalize by hand before passing the segments to the tokenizer I instantiate now (using `self.form_for_segmentation()`, just like in `self.tokenizer` seems like the most elegant way to do that). 

The `None` index in the dictionnary profile dict really threw me off, I thought I had found a bug before @tresoldi  explained that it is for multi-profile support. You are right,  @tresoldi, I will add a comment if I keep this way of instanciating the profile.

Commit:
Corrects test for alignment length, adds check for segmentation error symbol

Commit:
Changes orthography profile to keep valid alignments

Commit:
Adds all generated files after changes to profile an tokenization

XachaB commented:
I think everything should be done now. I kept the manually instantiated tokenizer and added normalization, but I can change it if using `self.tokenizer` is more in line with other repositories.

Note that the trick for tokenizing segments leads to dots in the generated Grapheme strings: `^ a . h . ts . i . a . m . a $`, but I am unsure whether this is a problem.

LinguList commented:
Yes, I think it is fine now. I do not mind the dots, they grapheme
strings are for transparency, so they reflect the hack. But maybe,
@xachab or @tresoldi, could you open a github issue on lexibank for
this, so we don't forget it: a) modifying already segmented data (best
practice with profiles, or without (?)), b) tests for formal
compatibility of alignments and segments (I will specify details, as
brackets are also used in edictor, etc.)

XachaB commented:
ok, I opened two issues

LinguList commented:
Thanks!

tresoldi commented:
I guess you can merge, @XachaB . :smiley:
